File: content/snippets/go/interview_lcci/08/12_eight_queens.go
package main

import (
	"fmt"
	"strings"
)

func solveNQueens(n int) [][]string {
	result := [][]string{}
	usedCols := map[int]bool{}
	diag1 := map[int]bool{}
	diag2 := map[int]bool{}
	board := make([][]byte, n)
	for i := range board {
		board[i] = []byte(strings.Repeat(".", n))
	}

	var bt func(row int)
	bt = func(row int) {
		if row == n {
			snapshot := make([]string, n)
			for i, r := range board {
				snapshot[i] = string(r)
			}
			result = append(result, snapshot)
			return
		}
		for col := 0; col < n; col++ {
			if usedCols[col] || diag1[row-col] || diag2[row+col] {
				continue
			}
			board[row][col] = 'Q'
			usedCols[col] = true
			diag1[row-col] = true
			diag2[row+col] = true
			bt(row + 1)
			board[row][col] = '.'
			delete(usedCols, col)
			delete(diag1, row-col)
			delete(diag2, row+col)
		}
	}

	bt(0)
	return result
}

func main() {
	result := solveNQueens(4)
	fmt.Printf("%d solutions\n", len(result))
	for _, board := range result {
		for _, row := range board {
			fmt.Println(row)
		}
		fmt.Println()
	}
}

File: content/snippets/cpp/interview_lcci/08/12_eight_queens.cpp
#include <iostream>
#include <vector>
#include <string>
#include <set>
#include <functional>
using namespace std;

vector<vector<string>> solveNQueens(int n) {
    vector<vector<string>> result;
    set<int> usedCols, diag1, diag2;

    function<void(int, vector<string>&)> bt = [&](int row, vector<string>& board) {
        if (row == n) { result.push_back(board); return; }
        for (int col = 0; col < n; col++) {
            if (usedCols.count(col) || diag1.count(row - col) || diag2.count(row + col)) continue;
            board[row][col] = 'Q';
            usedCols.insert(col); diag1.insert(row - col); diag2.insert(row + col);
            bt(row + 1, board);
            board[row][col] = '.';
            usedCols.erase(col); diag1.erase(row - col); diag2.erase(row + col);
        }
    };

    vector<string> board(n, string(n, '.'));
    bt(0, board);
    return result;
}

int main() {
    auto result = solveNQueens(4);
    cout << result.size() << " solutions" << endl;
    for (auto& board : result) {
        for (auto& row : board) cout << row << endl;
        cout << endl;
    }
    return 0;
}
